File: scripts/Partition List/Partition List.java
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode left = new ListNode(0);
        ListNode right = new ListNode(0);
        
        ListNode leftTail = left;
        ListNode rightTail = right;
        
        while(head != null){
            if(head.val < x){
                leftTail.next = head;
                leftTail = leftTail.next;
            }
            else{
                rightTail.next = head;
                rightTail = rightTail.next;
            }
            head = head.next;
        }
        
        leftTail.next = right.next;
        rightTail.next = null;
        
        return left.next;
    }
}

File: scripts/algorithms/P/Partition List/Partition List.rs
// Runtime: 0 ms (Top 100.0%) | Memory: 2.10 MB (Top 65.38%)

impl Solution {
    pub fn partition(mut head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let mut dummy1: Option<Box<ListNode>> = None;
        let mut dummy2: Option<Box<ListNode>> = None;
        let mut r1 = &mut dummy1;
        let mut r2 = &mut dummy2;
        while head.is_some() {
            if head.as_ref().unwrap().val < x {
                if r1.is_some() {
                    r1 = &mut r1.as_mut().unwrap().next;
                }
                std::mem::swap(r1, &mut head);
                std::mem::swap(&mut r1.as_mut().unwrap().next, &mut head);
            } else {
                std::mem::swap(r2, &mut head);
                std::mem::swap(&mut r2.as_mut().unwrap().next, &mut head);
                r2 = &mut r2.as_mut().unwrap().next;
            }
        }
        if r1.is_some() {
            r1.as_mut().unwrap().next = dummy2;
            dummy1
        } else {
            dummy2
        }
    }
}
